### ISSUE TITLE
bug: Aggregation field index off by one when joining two tables. 

### ISSUE THREAD

**Issue by mbwhite** [ibis-project/ibis-substrait#1072]
### What happened?

With the following query
```
        expr = (
            tests.inner_join(vehicles, vehicles.vehicle_id == tests.vehicle_id)        
            .group_by(vehicles.colour)
            .count()
            .select("colour", "CountStar()")
        )
```

This produces a plan; but the field offset used for the aggregation is 'off-by-one':   For example here's the plan created (I've a custom formatted for substrait)

```
<Substrait Plan>
Root::  ImmutableProject [colour, CountStar()]
<Substrait Relation>
+- Project:: [Str, I64, Str, I64]
   +- Aggregate:: FieldRef#/Str/StructField{offset=10}
      +- Join:: INNER <ScalarFn>equal:any_any
          : arg0 = FieldRef#/I32/StructField{offset=8}
          : arg1 = FieldRef#/I32/StructField{offset=1}
         +- NamedScan::  Tables=[tests] Fields=test_id[I32],vehicle_id[I32],test_date[Str],test_class[Str],test_type[Str],test_result[Str],test_mileage[I32],postcode_area[Str]
         +- NamedScan::  Tables=[vehicles] Fields=vehicle_id[I32],make[Str],model[Str],colour[Str],fuel_type[Str],cylinder_capacity[I32],first_use_date[Str]
```

Note that the offset of the aggregate is 10, and counting up that's `model` not `colour`;  this only occurs if there's a join, and only with the second table. So aggregating on say `test_type` would be ok. So I'm wondering if the issue is that the `vehicle_id` which is the join key is perhaps being ignored?

Can also be seen just printing out the protobuf

```
            groupings {
              grouping_expressions {
                selection {
                  direct_reference {
                    struct_field {
                      field: 10
                    }
                  }
                  root_reference {
                  }
                }
              }
            }
```

Specification of the tables is 
```
        tests = ibis.table(
            [
                ("test_id", "int32"),
                ("vehicle_id", "int32"),
                ("test_date", "str"),
                ("test_class", "str"),
                ("test_type", "str"),
                ("test_result", "str"),
                ("test_mileage", "int32"),
                ("postcode_area", "str"),
            ],
            name="tests",
        )

        vehicles = ibis.table(
            [
                ("vehicle_id", "int32"),
                ("make", "str"),
                ("model", "str"),
                ("colour", "str"),
                ("fuel_type", "str"),
                ("cylinder_capacity", "int32"),
                ("first_use_date", "str"),
            ],
            name="vehicles",
        )
```

I've looked at source code; but not spotted how this index is calculated for the aggregation case. 

### What version of ibis-substrait are you using?

v4

### What substrait consumer(s) are you using, if any?

Quite a few of them!  in this case I saw it via Spark; but the problem can been seen without a consumer

**Comment by gforsyth** [comment]
Thanks for the report @mbwhite and the easy reproducer!

Your diagnosis is correct -- the new `JoinChain` object deduplicates the join key, which is leading to the off-by-one error.  I'm taking a look at how to make sure we pull the field reference from the parent tables, instead.

**Comment by gforsyth** [comment]
Hey @mbwhite -- fix is merged and released as version 4.0.1 -- thanks again for flagging the issue!

**Comment by tokoko** [comment]
Hey, sorry for digging up an old ticket, but I was looking into this because of #1190.

@gforsyth At the time this ticket was created JoinChain was translated into just a series of JoinRels and ibis deduplicating a join key wasn't being taken into account leading to the bug described here. After the fix, we added handling of projects contained in joins (#1119) which was unrelated but also looks like inadvertently fixed this issue as well: JoinChain is now translated into a series of JoinRels + a final Project. This last projection node exposes the same schema as ibis JoinChain, therefore takes deduplication into account. If we take the example shown in the ticket, with the project node in place, offset of the aggregate should in fact be 10 now.

I'm going to remove the logic introduced by #1078 as it's no longer necessary and leading to other issues.
